### PR TITLE
Toggle timeline view's filter button to indicate excluded contributors (#145769)

### LINKED ISSUE #145769: Indicate filtered state on Timeline view
Currently there isn't any visual indication on the timeline view when a filter (i.e. anything unchecked from the list) has been applied:

CleanShot 2022-03-22 at 11 19 00@2x

One potential option here would be to follow the same dual icon pattern as pinning follows:

CleanShot 2022-03-22 at 11 20 56@2x

CleanShot 2022-03-22 at 11 21 04@2x

One other thought is that the timeline view header seems to indicated a pinned state with plain text as such. While this could be somewhat crammed to if it included filtered state as well, it might be worth exploring.

CleanShot 2022-03-22 at 11 21 37@2x

cc @bpasero @misolori

### PR DESCRIPTION
**Pull request by gjsjohnmurray**
This PR closes #145769 by allowing submenu-type action buttons to specify a `toggled` ContextKeyExpression that leverages existing infrastructure to use the `actionBar-toggledBackground` colour whenever the expression is true.

Here's how the Timeline view's Filter button now changes appearance when the user deselects at least one of the timeline contributors:

junk

The contrast of that colour in the Dark Modern theme maybe needs tweaking.

The recording also shows the highlighted filter button on the Problems tab of Panel, which was already using custom logic in its toolbar widget to achieve a similar effect. However it uses different colours:
https://github.com/microsoft/vscode/blob/d878a801801af60a91b8b50c16032386f40bdcc3/src/vs/workbench/browser/parts/views/media/views.css#L297-L301

Perhaps controversially, this PR had to increase the paneview header height from 22 to 23 pixels, otherwise the toggled button's background would overlap the focus rectangle on the header when focused. This happened because headers have a 1px top border to help separate them from the view above them, and this border pushes the action buttons down such that they overlap the bottom section of the focus rectangle when it appears.

### CODE REVIEW

**Comment by gjsjohnmurray** [comment]
Pinging @bpasero who contributed to the issue discussion.

**Comment by daviddossett** [comment]
Cool! Open question: should we use this subtle hover background-esque styling or the use the louder toggle state as shown above in the input? I'll actually pass this to @hbons since we was just exploring this recently.

File: src/vs/base/browser/ui/splitview/paneview.css [hunk 1]
```diff
@@ -21,7 +21,7 @@
 }
 
 .monaco-pane-view .pane > .pane-header {
-	height: 22px;
+	height: 23px;
```
Comment (bpasero) [inline src/vs/base/browser/ui/splitview/paneview.css:24]: We should not go away from our `22px` assumption in titles and headers. This looks like a hack to me.

File: src/vs/base/browser/ui/splitview/paneview.ts [hunk 2]
```diff
@@ -47,7 +47,7 @@ export interface IPaneStyles {
  */
 export abstract class Pane extends Disposable implements IView {
 
-	private static readonly HEADER_SIZE = 22;
+	private static readonly HEADER_SIZE = 23;
```
Comment (bpasero) [inline src/vs/base/browser/ui/splitview/paneview.ts:50]: We should not go away from our `22px` assumption in titles and headers.

**Review by bpasero — CHANGES_REQUESTED** [review]

**Comment by bpasero** [comment]
Adding some more people as this change seems to touch different areas. I only observe that going from `22px` to `23px` seems wrong to me and will result in inconsistent height assumptions. `22px` is our standard used throughout all of the workbench.

File: src/vs/platform/actions/browser/menuEntryActionViewItem.ts [hunk 3]
```diff
@@ -317,6 +318,10 @@ export class SubmenuEntryActionViewItem extends DropdownMenuActionViewItem {
 
 		container.classList.add('menu-entry');
 		const action = <SubmenuItemAction>this._action;
+		const toggled: ContextKeyExpression | undefined = action.item.toggled;
+		if (toggled && toggled.evaluate(this._contextKeyService.getContext(container))) {
```
Comment (jrieken) [inline src/vs/platform/actions/browser/menuEntryActionViewItem.ts:322]: This shouldn't be here. The action is just displaying/representing its properties and shouldn't evaluate/recompute ones. There is the toggled/checked property on the action which could be used but TBH I have concerns (see below)

File: src/vs/platform/actions/common/actions.ts [hunk 4]
```diff
@@ -33,6 +33,7 @@ export interface ISubmenuItem {
 	order?: number;
 	isSelection?: boolean;
 	rememberDefaultAction?: boolean;	// for dropdown menu: if true the last executed action is remembered as the default action
+	toggled?: ContextKeyExpression;
```
Comment (jrieken) [inline src/vs/platform/actions/common/actions.ts:36]: I doubt that a submenu should be toggled. What should that mean in the domain of actions, menus and submenus? Submenus are just containers for other menus. I do understand the use-case that's tackled here but I believe that's UI only. E.g, the rendering of a submenu should check its actions and render accordingly but it shouldn't have a toggled properties itself.

**Review by jrieken — CHANGES_REQUESTED** [review]

File: src/vs/platform/actions/common/actions.ts (hunk 4 above)
Comment (jrieken) [reply src/vs/platform/actions/common/actions.ts:36]: e.g [SubmenuEntryActionViewItem](https://github.com/microsoft/vscode/blob/fa45b00051606aa4e7bd8b4da05f209faa5ba02d/src/vs/platform/actions/browser/menuEntryActionViewItem.ts#L295) is IMO the better place to implement this, without new model properties/concept but purely rendering based

**Comment by hbons** [comment]
It looks a bit weird to me that the toggled state appears on buttons that only show up on hover... If there is a change of state that needs to be reflected at all times we should make that button always visible?

**Comment by gjsjohnmurray** [comment]
I reverted the 23px changes. Here's an image of the problem caused by the 1px border-top added here when the view is anywhere other than at the top of the primary sidebar container:

https://github.com/microsoft/vscode/blob/d7ce6f942494744686fcfb258bdb4b2ecfb526dc/src/vs/base/browser/ui/splitview/paneview.ts#L309

image

**Comment by gjsjohnmurray** [comment]
Closing this and replacing it with #198979 as a result of the feedback from @jrieken

**Pull request closed without merging**